### PR TITLE
Add Consul module

### PR DESCRIPTION
**Pull request by julb**

### CODE REVIEW

**Comment by kiview** [comment]
Hi @julb, thanks a lot for contributing this module.

Looking at the logic around `runInitCommands()`, it seems that this module makes using Consul with Testcontainers considerably easier, so theoretically we can imagine this becoming an incubating module.

However, we are currently lacking the capacity to review and merge PRs for new modules, so I can't promise when we will be able to get to it. This does not mean, that we don't appreciate your contribution. We are currently in the process to rework our current approach around modules, to make it easier for the community to contribute them.

In case you are interested in getting it out to users as early as possible, you can always consider publishing it in your own repo (this is in no way an indication, that we don't want to add it to our repo).

File: docs/modules/consul.md
Comment (eddumelendez) [inline docs/modules/consul.md:36]: use snippet from test code. See other modules as a reference.

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulClientTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulClientTest.java:18]: we can move this test to ConsultContainerTest

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:69]: this is using a similar approach from vault. I would suggest to remove `addProperties` and rename `runInitCommands` with `runConsulCommands` with a general purpose.

`addProperties` is pretty convenient method to expose but it will couple the container class to one specific command from consul.

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:46]: remove it

**Review by eddumelendez — CHANGES_REQUESTED** [review]
Hi @julb, sorry for the late reply and thanks for your contribution! The code looks good but I've added a few comments.

Are you willing to update the PR?

**Comment by julb** [comment]
Hey @eddumelendez 
It is updated.
Thanks for your feedback!

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:22]: update docs accordingly

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:87]: withConsulCommand?

**Review by eddumelendez — COMMENTED** [review]
just a couple of comments and LGTM. Once again, thanks!

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:23]:
```suggestion
public class ConsulContainer extends GenericContainer<ConsulContainer> {
```

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:33]: drop them

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:46]:
```suggestion
        setWaitStrategy(Wait.forHttp("/v1/status/leader").forPort(CONSUL_HTTP_PORT).forStatusCode(200));
```

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:49]:
```suggestion
        withEnv("CONSUL_ADDR", "http://0.0.0.0:" + CONSUL_HTTP_PORT);
```

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (eddumelendez) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:51]:
```suggestion
        withExposedPorts(CONSUL_HTTP_PORT, CONSUL_GRPC_PORT);
```

**Review by eddumelendez — CHANGES_REQUESTED** [review]
few more comments from @kiview and me :)

**Comment by julb** [comment]
@eddumelendez all remarks addressed! Thanks a lot.

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java:32]: I think `withConsulCommand` can be used here too

```suggestion
        GenericContainer.ExecResult result = consulContainer.withConsulCommand("kv get config/testing1");
```

**Review by eddumelendez — APPROVED** [review]
LGTM

File: modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java
Comment (kiview) [inline modules/consul/src/main/java/org/testcontainers/consul/ConsulContainer.java:19]:
```suggestion
 * Testcontainers implementation for Consul.
```

**Review by kiview — APPROVED** [review]
Thanks a lot for the PR @julb, LGTM 👍 

There is one open question regarding the package name, but this is something we as maintainer should align on first 🙂

**Comment by eddumelendez** [comment]
@julb can you please add consult to

* .github/ISSUE_TEMPLATE/bug_report.yaml
* .github/ISSUE_TEMPLATE/enhancement.yaml
* .github/ISSUE_TEMPLATE/feature.yaml
* .github/dependabot.yml
* .github/labeler.yml

```diff
Index: .github/ISSUE_TEMPLATE/bug_report.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/ISSUE_TEMPLATE/bug_report.yaml b/.github/ISSUE_TEMPLATE/bug_report.yaml
--- a/.github/ISSUE_TEMPLATE/bug_report.yaml	(revision c9fc9afff3c13bdf7c6f7762ab8815ed91550399)
+++ b/.github/ISSUE_TEMPLATE/bug_report.yaml	(revision 95124487415f8ffad3a06df54d6081f4dd6e269d)
@@ -17,6 +17,7 @@
         - Azure
         - Cassandra
         - Clickhouse
+        - Consul
         - Couchbase
         - DB2
         - Dynalite
Index: .github/ISSUE_TEMPLATE/enhancement.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/ISSUE_TEMPLATE/enhancement.yaml b/.github/ISSUE_TEMPLATE/enhancement.yaml
--- a/.github/ISSUE_TEMPLATE/enhancement.yaml	(revision c9fc9afff3c13bdf7c6f7762ab8815ed91550399)
+++ b/.github/ISSUE_TEMPLATE/enhancement.yaml	(revision 95124487415f8ffad3a06df54d6081f4dd6e269d)
@@ -17,6 +17,7 @@
         - Azure
         - Cassandra
         - Clickhouse
+        - Consul
         - Couchbase
         - DB2
         - Dynalite
Index: .github/ISSUE_TEMPLATE/feature.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/ISSUE_TEMPLATE/feature.yaml b/.github/ISSUE_TEMPLATE/feature.yaml
--- a/.github/ISSUE_TEMPLATE/feature.yaml	(revision c9fc9afff3c13bdf7c6f7762ab8815ed91550399)
+++ b/.github/ISSUE_TEMPLATE/feature.yaml	(revision 95124487415f8ffad3a06df54d6081f4dd6e269d)
@@ -17,6 +17,7 @@
         - Azure
         - Cassandra
         - Clickhouse
+        - Consul
         - Couchbase
         - DB2
         - Dynalite
Index: .github/dependabot.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/dependabot.yml b/.github/dependabot.yml
--- a/.github/dependabot.yml	(revision c9fc9afff3c13bdf7c6f7762ab8815ed91550399)
+++ b/.github/dependabot.yml	(revision 95124487415f8ffad3a06df54d6081f4dd6e269d)
@@ -43,6 +43,11 @@
       interval: "monthly"
     open-pull-requests-limit: 10
   - package-ecosystem: "gradle"
+    directory: "/modules/consul"
+    schedule:
+      interval: "monthly"
+    open-pull-requests-limit: 10
+  - package-ecosystem: "gradle"
     directory: "/modules/couchbase"
     schedule:
       interval: "monthly"
Index: .github/labeler.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/labeler.yml b/.github/labeler.yml
--- a/.github/labeler.yml	(revision c9fc9afff3c13bdf7c6f7762ab8815ed91550399)
+++ b/.github/labeler.yml	(revision 95124487415f8ffad3a06df54d6081f4dd6e269d)
@@ -15,6 +15,8 @@
   - modules/clickhouse/*
 "modules/cockroachdb":
   - modules/cockroachdb/*
+"modules/consul":
+  - modules/consul/*
 "modules/couchbase":
   - modules/couchbase/*
 "modules/db2":

```

**Comment by julb** [comment]
@eddumelendez @kiview done!
Thanks for your reviews and your time.

**Comment by kiview** [comment]
Thanks @julb, we will merge once we internally decided on how to proceed with package names for container modules 🙂

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java:65]: import static

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java:34]: move to assertThat from assertj, please

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java:71]: import static

File: modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java
Comment (eddumelendez) [inline modules/consul/src/test/java/org/testcontainers/consul/ConsulContainerTest.java:48]: can we do something like

```suggestion
        io.restassured.response.Response response = RestAssured
            .given()
            .when()
            .get("http://" + getHostAndPort() + "/v1/kv/config/testing1")
            .andReturn();
        assertThat(response.body().jsonPath().getString("[0].Value"))
            .isEqualTo(Base64.getEncoder().encodeToString("value123".getBytes(StandardCharsets.UTF_8)));
```

**Review by eddumelendez — COMMENTED** [review]
one last thing @julb ! We are not able to push to your branch so here we are again asking you for some changes 😬 about moving to `assertj`

Thank you so much!

**Comment by eddumelendez** [comment]
Thanks @julb ! PR has been merged and now we have a consul module!

**Comment by julb** [comment]
OMG!
Thanks a lot for your support, that's awesome !!

**Pull request merged**
